### PR TITLE
feat(dart/transform): Populate `lifecycle` from lifecycle interfaces

### LINKED ISSUE #3181: directive_metadata_reader should take lifecycle interfaces into account
Right now, directive_metadata_reader will not correctly populate its lifecycle booleans when lifecycle interfaces are implemented by a `Directive`.

The result is that interface-based lifecycle methods work in mirror mode but fail in transformed mode.

The workaround is to explicitly declare lifecycle methods in the `lifecycle` parameter of the `Directive`.

### PR DESCRIPTION
**Pull request by kegluneq**
When a `Directive` implements or extends a lifecycle interface (e.g. `OnChange` or
`OnInit`), populate its `lifecycle` property if not already populated).

Closes #3181

### CODE REVIEW

**Comment by kegluneq** [comment]
/cc @TedSander , this will remove the need to explicitly specify `lifecycle` values for transformed outputs.

I'd like to have @jakemac53 review since the `InterfaceMatcher` is very similar to (and reuses much of) his `AnnotationMatcher` class.

**Comment by TedSander** [comment]
SGTM

File: modules/angular2/src/transform/common/class_matcher_base.dart [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+  /// Checks whether an [Identifier] matches any [ClassDescriptor].
+  ImportDirective getMatchingImport(Identifier className, AssetId assetId) {
+    for (var d in _classDescriptors) {
+      var matchingImport = _getMatchingImport(className, d, assetId);
+      if (matchingImport != null) {
+        return matchingImport;
+      }
+    }
+    return null;
```
Comment (jakemac53) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: not necessary afaik

File: modules/angular2/src/transform/common/interface_matcher.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+import 'package:barback/barback.dart' show AssetId;
+import 'class_matcher_base.dart';
+
+export 'class_matcher_base.dart' show ClassDescriptor;
+
+/// [ClassDescriptor]s for the default angular interfaces that may be
+/// implemented by a class. These classes are re-exported in many places so this
+/// covers all libraries which provide them.
+const _ON_CHANGE_INTERFACES = const [
+  const ClassDescriptor('OnChange', 'package:angular2/annotations.dart', null),
```
Comment (jakemac53) [inline modules/angular2/src/transform/common/interface_matcher.dart]: These should also probably all include `package:angular2/angular2.dart`.

File: modules/angular2/src/transform/common/class_matcher_base.dart [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    Identifier className, ClassDescriptor descriptor, AssetId assetId) {
+  return _getMatchingImport(className, descriptor, assetId) != null;
+}
+
+/// String based description of a class and its location.
+class ClassDescriptor {
+  /// The name of the class.
+  final String name;
+  /// A `package:` style import path to the file where the class is defined.
+  final String import;
```
Comment (jakemac53) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: It is probably worth converting this to a list? Could be a separate PR though.

File: modules/angular2/src/transform/common/class_matcher_base.dart [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  /// The class that this class extends or implements. This is the only optional
+  /// field.
+  final String superClass;
+
+  AssetId get assetId => new AssetId(package, packagePath);
+  String get package => path.split(import.replaceFirst('package:', '')).first;
+  String get packagePath => path.joinAll(['lib']
+    ..addAll(path.split(import.replaceFirst('package:', ''))..removeAt(0)));
+
+  const ClassDescriptor(this.name, this.import, this.superClass);
```
Comment (jakemac53) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: Lets also make this.superClass an optional arg

File: modules/angular2/src/transform/directive_processor/rewriter.dart [hunk 5]
```diff
@@ -22,13 +23,15 @@ import 'visitors.dart';
...
-    AnnotationMatcher annotationMatcher, bool inlineViews) async {
+Future<String> createNgDeps(
+    AssetReader reader, AssetId assetId, AnnotationMatcher annotationMatcher,
+    {bool inlineViews}) async {
   // TODO(kegluneq): Shortcut if we can determine that there are no
   // [Directive]s present, taking into account `export`s.
   var writer = new AsyncStringWriter();
   var visitor = new CreateNgDepsVisitor(writer, assetId,
-      new XhrImpl(reader, assetId), annotationMatcher, inlineViews);
+      new XhrImpl(reader, assetId), annotationMatcher, new InterfaceMatcher(),
```
Comment (jakemac53) [inline modules/angular2/src/transform/directive_processor/rewriter.dart]: Probably worth making a single top level `InterfaceMatcher` for this. It creates a fair number of `ClassDescriptor`s in its constructor so its somewhat expensive.

File: modules/angular2/src/transform/directive_processor/visitors.dart [hunk 6]
```diff
@@ -203,25 +206,83 @@ class FactoryTransformVisitor extends _CtorTransformVisitor {
...
-      AsyncStringWriter writer, this._xhr, this._inlineViews)
+  AnnotationsTransformVisitor(AsyncStringWriter writer, this._xhr,
+      this._annotationMatcher, this._interfaceMatcher, this._assetId,
+      {bool inlineViews})
       : this.writer = writer,
+        _inlineViews = inlineViews,
         super(writer);
 
+  /// Determines if the `node` has interface-based lifecycle methods and populates
+  /// `_lifecycleValue` with the appropriate values if so. If none are present, `_lifecycleValue` is not modified.
```
Comment (jakemac53) [inline modules/angular2/src/transform/directive_processor/visitors.dart]: nit: line length

File: modules/angular2/src/transform/directive_processor/visitors.dart [hunk 7]
```diff
@@ -203,25 +206,83 @@ class FactoryTransformVisitor extends _CtorTransformVisitor {
...
+  void _populateLifecycleValue(ClassDeclaration node) {
+    var lifecycleEntries = [];
+    var prefix = '';
+    var populateImport = (Identifier name) {
+      if (prefix.isNotEmpty) return;
+      var import = _interfaceMatcher.getMatchingImport(name, _assetId);
+      prefix =
+      import != null && import.prefix != null ? '${import.prefix}.' : '';
```
Comment (jakemac53) [inline modules/angular2/src/transform/directive_processor/visitors.dart]: nit: format

File: modules/angular2/src/transform/common/class_matcher_base.dart (hunk 1 above)
Comment (kegluneq) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: Removed

File: modules/angular2/src/transform/common/interface_matcher.dart (hunk 2 above)
Comment (kegluneq) [inline modules/angular2/src/transform/common/interface_matcher.dart]: Done

File: modules/angular2/src/transform/common/class_matcher_base.dart (hunk 3 above)
Comment (kegluneq) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: Good call on both counts -- probably worth it, and probably worth doing in a separate PR

File: modules/angular2/src/transform/common/class_matcher_base.dart (hunk 4 above)
Comment (kegluneq) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: I think named is cleanest here, what do you think?

File: modules/angular2/src/transform/directive_processor/rewriter.dart (hunk 5 above)
Comment (kegluneq) [inline modules/angular2/src/transform/directive_processor/rewriter.dart]: Done

File: modules/angular2/src/transform/directive_processor/visitors.dart (hunk 6 above)
Comment (kegluneq) [inline modules/angular2/src/transform/directive_processor/visitors.dart]: Done

File: modules/angular2/src/transform/common/class_matcher_base.dart (hunk 4 above)
Comment (jakemac53) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: sure

File: modules/angular2/test/transform/directive_processor/interface_lifecycle_files/expected/soup.ng_deps.dart [hunk 8]
```diff
@@ -0,0 +1,59 @@
+library dinner.soup.ng_deps.dart;
```
Comment (jakemac53) [inline modules/angular2/test/transform/directive_processor/interface_lifecycle_files/expected/soup.ng_deps.dart]: It would be good to have one component that implements more than one interface as well, to verify the const list construction

Comment (kegluneq) [inline modules/angular2/test/transform/directive_processor/interface_lifecycle_files/expected/soup.ng_deps.dart]: Great idea, added.

File: modules/angular2/src/transform/common/class_matcher_base.dart (hunk 1 above)
Comment (jakemac53) [inline modules/angular2/src/transform/common/class_matcher_base.dart]: kk
